Index: Makefile
===================================================================
--- Makefile	(revision 31028)
+++ Makefile	(working copy)
@@ -618,6 +618,7 @@
 SRCS_MPLAYER-$(LIBMENU_DVBIN) += libmenu/menu_dvbin.c
 SRCS_MPLAYER-$(LIRC)          += input/lirc.c
 SRCS_MPLAYER-$(MD5SUM)        += libvo/vo_md5sum.c
+SRCS_MPLAYER-$(DMX)           += libvo/vo_dmx.c
 SRCS_MPLAYER-$(MGA)           += libvo/vo_mga.c
 SRCS_MPLAYER-$(NAS)           += libao2/ao_nas.c
 SRCS_MPLAYER-$(OPENAL)        += libao2/ao_openal.c
Index: libvo/vo_dmx.c
===================================================================
--- libvo/vo_dmx.c	(revision 0)
+++ libvo/vo_dmx.c	(revision 0)
@@ -0,0 +1,184 @@
+/*
+ * This file is part of MPlayer.
+ *
+ * MPlayer is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * MPlayer is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with MPlayer; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <math.h>
+
+#include "config.h"
+#include "mp_msg.h"
+#include "help_mp.h"
+#include "video_out.h"
+#include "video_out_internal.h"
+#include "libswscale/swscale.h"
+#include "libmpcodecs/vf_scale.h"
+
+#include <cdmx/plugin.h>
+
+static const vo_info_t info =
+{
+	"DMX Output",
+	"dmx",
+	"Michael Axiak <axiak@mit.edu>",
+	""
+};
+
+
+const LIBVO_EXTERN (dmx)
+
+static uint8_t *slice_data;
+static int img_height;
+static int img_width;
+
+/* DMX stuff */
+static LocalData *s;
+
+/* used for the sws */
+static struct SwsContext *sws=NULL;
+static uint8_t * image[MP_MAX_PLANES];
+static int image_stride[MP_MAX_PLANES];
+static uint32_t imgformat;
+
+static uint32_t draw_image(mp_image_t* mpi)
+{
+    uint8_t * rgbdata;
+    register int baseidx, width;
+
+    if(mpi->flags&(MP_IMGFLAG_DIRECT|MP_IMGFLAG_DRAW_CALLBACK)) return VO_TRUE;
+
+    if (!sws) {
+        sws = sws_getContextFromCmdLine(mpi->w, mpi->h, imgformat,
+                                        s->layer->width, s->layer->height,
+                                        IMGFMT_RGB24);
+        memset(image_stride, 0, sizeof(image_stride));
+        for (int i=0; i < MP_MAX_PLANES; i++) {
+            image[i] = (uint8_t *)malloc(3 * s->layer->width * s->layer->height);
+        }
+        image_stride[0] = 3 * s->layer->width;
+    }
+
+    sws_scale(sws, mpi->planes, mpi->stride, 0, mpi->h, image, image_stride);
+
+    width = s->layer->width;
+    rgbdata = image[0];
+    for (int r = 0; r < s->layer->height; r++) {
+        for (int c = 0; c < s->layer->width; c++) {
+            baseidx = 3*(c + width * r);
+            rgbpixel_setintvalue(
+                                 colorlayer_getpixel(s->layer, c,
+                                                     s->layer->height - r - 1),
+                                 rgbdata[baseidx],
+                                 rgbdata[baseidx + 1],
+                                 rgbdata[baseidx + 2],
+                                 255);
+        }
+    }
+
+    serverdata_commitlayer(s);
+
+    return VO_TRUE;
+}
+
+static int config(uint32_t width, uint32_t height, uint32_t d_width, uint32_t d_height, uint32_t flags, char *title, uint32_t format)
+{
+    mp_msg(MSGT_VO, MSGL_V, "DMX: Configure entered.\n");
+
+    img_width = width;
+    img_height = height;
+
+    imgformat = format;
+
+    if(sws) sws_freeContext(sws);
+    for (int i=0; i < MP_MAX_PLANES; i++) {
+        if (image[i]) free(image[i]);
+    }
+    return 0;
+}
+
+static void draw_osd(void)
+{
+}
+
+static void flip_page (void)
+{
+    return;
+}
+
+static int draw_slice(uint8_t *srcimg[], int stride[], int w,int h,int x,int y)
+{
+    return -1;
+}
+
+static int draw_frame(uint8_t * src[])
+{
+    return -1;
+}
+
+static int query_format(uint32_t format)
+{
+    switch(format){
+	case IMGFMT_YV12:
+	case IMGFMT_I420:
+	case IMGFMT_IYUV:
+	case IMGFMT_IYU2:
+	case IMGFMT_BGR32:
+	case IMGFMT_BGR24:
+	case IMGFMT_BGR16:
+	case IMGFMT_BGR15:
+	case IMGFMT_RGB32:
+	case IMGFMT_RGB24:
+	case IMGFMT_Y8:
+	case IMGFMT_Y800:
+	    return VFCAP_CSP_SUPPORTED | VFCAP_SWSCALE | VOCAP_NOSLICES;
+    }
+    return 0;
+}
+
+static void uninit(void)
+{
+    if (slice_data != NULL) free(slice_data);
+    if(sws) sws_freeContext(sws);
+}
+
+static void check_events(void)
+{
+}
+
+static int preinit(const char *arg)
+{
+    s = plugin_register("vo_dmx", 155);
+    if (!s) {
+        mp_msg(MSGT_VO, MSGL_ERR, "vo_dmx: Couldn't initialize DMX system\n");
+        return -1;
+    }
+    return 0;
+}
+
+static int control(uint32_t request, void *data, ...)
+{
+  switch (request) {
+      case VOCTRL_DRAW_IMAGE:
+          return draw_image(data);
+
+      case VOCTRL_QUERY_FORMAT:
+          return query_format(*((uint32_t*)data));
+  }
+  return VO_NOTIMPL;
+}
Index: libvo/video_out.c
===================================================================
--- libvo/video_out.c	(revision 31028)
+++ libvo/video_out.c	(working copy)
@@ -139,6 +139,7 @@
 extern const vo_functions_t video_out_quartz;
 extern const vo_functions_t video_out_pnm;
 extern const vo_functions_t video_out_md5sum;
+extern const vo_functions_t video_out_dmx;
 
 const vo_functions_t* const video_out_drivers[] =
 {
@@ -287,6 +288,9 @@
 #ifdef CONFIG_MD5SUM
         &video_out_md5sum,
 #endif
+#ifdef CONFIG_DMX
+        &video_out_dmx,
+#endif
         NULL
 };
 
Index: configure
===================================================================
--- configure	(revision 31028)
+++ configure	(working copy)
@@ -375,6 +375,7 @@
   --enable-ggi             enable GGI video output [autodetect]
   --enable-ggiwmh          enable GGI libggiwmh extension [autodetect]
   --enable-direct3d        enable Direct3D video output [autodetect]
+  --enable-dmx             enable DMX video output [autodetect]
   --enable-directx         enable DirectX video output [autodetect]
   --enable-dxr2            enable DXR2 video output [autodetect]
   --enable-dxr3            enable DXR3/H+ video output [autodetect]
@@ -581,6 +582,7 @@
 _sdl=auto
 _kva=auto
 _direct3d=auto
+_dmx=auto
 _directx=auto
 _win32waveout=auto
 _nas=auto
@@ -926,6 +928,8 @@
   --disable-kva)        _kva=no         ;;
   --enable-direct3d)    _direct3d=yes   ;;
   --disable-direct3d)   _direct3d=no    ;;
+  --enable-dmx)         _dmx=yes        ;;
+  --disable-dmx)        _dmx=no         ;;
   --enable-directx)     _directx=yes    ;;
   --disable-directx)    _directx=no     ;;
   --enable-win32waveout)  _win32waveout=yes ;;
@@ -4314,6 +4318,32 @@
 echores "$_bl"
 
 
+echocheck "dmx"
+if test "$_dmx" = yes ; then
+    def_dmx="#define CONFIG_DMX 1"
+    _ld_dmx="-ldmxplugin"
+    libs_mplayer="$libs_mplayer $_ld_dmx"
+    _vomodules="dmx $_vomodules"
+fi
+if test "$_dmx" = auto ; then
+    _dmx=no
+  cat > $TMPC <<EOF
+#include <cdmx/plugin.h>
+int main(void) { return 0; }
+EOF
+
+  cc_check -ldmxplugin && _dmx=yes && \
+      def_dmx="#define CONFIG_DMX 1" && \
+      libs_mplayer="$libs_mplayer -ldmxplugin" && \ 
+  _vomodules="dmx $_vomodules"
+  
+  cc_check $_ld_tmp && libs_mplayer="$libs_mplayer $_ld_tmp" \
+      && _x11=yes && break
+
+fi
+echores "$_dmx"
+
+
 echocheck "DirectFB"
 if test "$_directfb" = auto ; then
   _directfb=no
@@ -8520,6 +8550,7 @@
 DIRECT3D = $_direct3d
 DIRECTFB = $_directfb
 DIRECTX = $_directx
+DMX = $_dmx
 DVBIN = $_dvbin
 DVDNAV = $_dvdnav
 DVDNAV_INTERNAL = $dvdnav_internal
@@ -9090,6 +9121,7 @@
 $def_jpeg
 $def_kva
 $def_md5sum
+$def_dmx
 $def_mga
 $def_mng
 $def_png
